src/porepy/numerics/time_step_control.py

```diff
@@ -1,130 +1,191 @@
-        elif dt_min_max[1] * iter_lowupp_factor[1] < dt_min_max[0]:
-            raise ValueError("Encountered dt_max * iter_upp_factor < dt_min.")
-
-        # Check for recomputation factor
-        if recomp_factor >= 1:
-            raise ValueError("Expected recomputation multiplication factor < 1.")
-
-        # Check for maximum number of recomputation attempts
-        if recomp_max <= 0:
-            raise ValueError("Number of recomputation attempts must be > 0.")
+        # Set dt_min_max if necessary
+        if dt_min_max is None:
+            dt_min_max = (0.001 * schedule[-1], 0.1 * schedule[-1])
+
+        # More sanity checks below. Note that all the remaining sanity checks (but one) are
+        # only needed when constant_dt = False. Thus, to save time when constant_dt = True,
+        # we use this rather ugly if-statement from below.
+
+        if not constant_dt:
+
+            # Sanity checks for dt_min and dt_max
+            if dt_init < dt_min_max[0]:
+                raise ValueError(
+                    "Initial time step cannot be smaller than minimum time step."
+                )
+            elif dt_init > dt_min_max[1]:
+                raise ValueError(
+                    "Initial time step cannot be larger than maximum time step."
+                )
+
+            # NOTE: The above checks guarantee that minimum time step <= maximum time step
+
+            # Sanity checks for maximum number of iterations.
+            # Note that 0 is a possibility. This will imply that the solver reaches
+            # convergence directly, e.g., as in linear solvers.
+            if iter_max < 0:
+                raise ValueError("Maximum number of iterations must be non-negative.")
+
+            # Sanity checks for optimal iteration range
+            if iter_optimal_range[0] > iter_optimal_range[1]:
+                s = "Lower optimal iteration range cannot be larger than"
+                s += " upper optimal iteration range."
+                raise ValueError(s)
+            elif iter_optimal_range[1] > iter_max:
+                s = "Upper optimal iteration range cannot be larger than"
+                s += " maximum number of iterations."
+                raise ValueError(s)
+            elif iter_optimal_range[0] < 0:
+                s = "Lower optimal iteration range cannot be negative."
+                raise ValueError(s)
+
+            # Sanity checks for lower and upper multiplication factors
+            if iter_lowupp_factor[0] <= 1:
+                raise ValueError("Expected lower multiplication factor > 1.")
+            elif iter_lowupp_factor[1] >= 1:
+                raise ValueError("Expected upper multiplication factor < 1.")
+
+            # Checks for sensible combinations of iter_optimal_range and iter_lowupp_factor
+            if dt_min_max[0] * iter_lowupp_factor[0] > dt_min_max[1]:
+                raise ValueError("Encountered dt_min * iter_low_factor > dt_max.")
```

Comment:
I understand the logic here, but I did not expected such requirements on the parameters in addition to what was described above (in particular since the documentation is so good, well done!). My suggestion is to give the update formula in the class documentation, and then write short comments just above these error messages so that a user would understand what went wrong. You could also document this in the class documentation, but that would lead to  so many conditions that it is easy to overlook something.

Reply:
I'm quite sure that in neither of these cases the algorithm will crash. In the first case, it will simply select `dt_max`, and in the second case `dt_min`.

For the record, I've added these checks because when such choices of `dt_min_max` and `iter_lowupp_factor` are made, the algorithm can't really do much (except selecting `dt_max` and `dt_min`). Nevertheless, this does not qualify as an error, but as a (very bad) decision from the user side. And we cannot, of course, cover the whole set of bad choices (believe me, it's a huge set). 

That being said, and to your point, I would suggest demoting this error to a (more meaningful) warning, provided the algorithm does not fail. This, we will have to make sure with unit tests. If it turns out that the algorithm actually fail, we'll go with your suggestion, since, again, this is really a more or less edge case. How does that sound?

Reply:
First impression is that I agree on the warning suggestion. But could this kind of incompatible values lead to bad behaviour? E.g. first relaxing the time step, and then restricting because the restriction rendered it larger than dt_max?

Reply:
Indeed, this oscillatory behavior that you describe might be a possibility for really bad choices of parameters. But in principle, it won't break the algorithm. It will most likely jump from `dt_min` to `dt_max` between iteration steps. 

We have to decide whether we want to allow this or not.

Reply:
My opinion:
1) We want to check for this, users (including myself) can easily mix this up, and it is hard to debug.
2) Add sufficient documentation so that the user does not get an error message from what seems to be acceptable parameter values.

Reply:
Ok. Then it stays as an error, but with a more complete error message.

Reply:
I've added extensive documentation.